[Structured Kernels] Port for `baddbmm` and `bmm`

krshrimali opened the pull request:
This PR attempts to port `baddbmm` and `bmm` to structured kernels. The reason it's in the same PR: because a lot of it is common for both the ops, including the checks and implementation.

Issue tracker: https://github.com/pytorch/pytorch/issues/55070

cc: @ysiraichi @ezyang

krshrimali committed:
WIP, bmm and baddbmm

krshrimali committed:
Init: bmm and baddbmm

krshrimali committed:
Almost there, lots of const_casts :/

krshrimali committed:
This now builds, but breaks everything else :)

krshrimali committed:
One last failure to fix, and then done...

krshrimali committed:
Fix final failures, all tests pass on CPU

krshrimali committed:
I might just not move checks from cuda...

krshrimali committed:
Final fixes

krshrimali committed:
Remove commented out code

krshrimali committed:
Remove resize_output from cuda code

krshrimali committed:
Remove import (earlier used for debugging)

krshrimali committed:
Remove whitespaces

krshrimali committed:
Whitespaces removed

krshrimali committed:
Typo fix in the error msg

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into structured_kernels/bmm

krshrimali commented:
A few failures are there, working on them. I will make it ready for review once they are resolved.

krshrimali committed:
Fix for named tensors

krshrimali committed:
Merge branch 'structured_kernels/bmm' of github.com:krshrimali/pytorch into structured_kernels/bmm

krshrimali committed:
Replace bmm calls

krshrimali committed:
Minor fix

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,86 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+void common_checks_baddbmm_bmm(Meta& meta, const Tensor& batch1, const Tensor& batch2, const Scalar& beta, const Scalar& alpha, bool is_bmm, const c10::optional<Tensor>& self_or_result=nullopt) {
+  CheckedFrom c = (is_bmm ? "bmm" : "baddbmm");
+
+  auto checkOnCPU = [](const Tensor& t, CheckedFrom c) {
+    TORCH_CHECK(
+      !t.is_cuda(),
+      "Expect tensor to have CPU backend, but got tensor with ",
+      toString(t.options().backend()),
+      " Backend (while checking arguments for ",
+      c);
+  };
+
+  if (self_or_result.has_value()) checkOnCPU(self_or_result.value(), c);
+  checkOnCPU(batch1, c);
+  checkOnCPU(batch2, c);
```
I think checking whether tensors are in the same device is already taken care of because of how dispatch works (given that this function is only called for CPU, CUDA, and META backends).

krshrimali replied:
I think these checks were useful since the implementations are different, so just to make sure no one in future calls this impl for CUDA. I've removed these checks from the meta function, but have kept them for impl, for consistency and correctness in the future.

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,86 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+  checkOnCPU(batch1, c);
+  checkOnCPU(batch2, c);
+
+  checkDim(c, batch1, "batch1", /* pos */1, /* dim */ 3);
+  checkDim(c, batch2, "batch2", /* pos */ 2, /* dim */ 3);
```
These 2 checks are the same as the ones for CUDA in the `META` function, aren't they? Aren't they device-independent?

krshrimali replied:
Right, indeed they are! This is resolved now.

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,86 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+  int64_t res_rows = batch1_sizes[1];
+  int64_t res_cols = batch2_sizes[2];
+
+  TORCH_CHECK(batch2_sizes[0] == bs && batch2_sizes[1] == contraction_size);
```
Device-independent check?

krshrimali replied:
Indeed, thanks for pointing this out.

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,86 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+    meta.set_output({bs, res_rows, res_cols}, batch1.options());
+  } else {
+    if (self_or_result.has_value()) {
+      const auto self_sizes = self_or_result.value().sizes();
+      TORCH_CHECK(self_sizes[0] == bs && self_sizes[1] == res_rows && self_sizes[2] == res_cols);
```
Device-independent check?

krshrimali replied:
Actually, yes, you are right.

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,86 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+    if (self.is_cuda()) {
+        auto& result = maybe_get_output(0);
+
+        TORCH_CHECK(self.dim() == 3, "batch1 must be a 3D tensor");
+        TORCH_CHECK(mat2.dim() == 3, "batch2 must be a 3D tensor");
+
+        set_output({self.sizes()[0], self.sizes()[1], mat2.sizes()[2]}, self.options());
```
Device-independent checks?

krshrimali replied:
Fixed now

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1238,12 +1318,14 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
 
```
Aren't the lines above (until the beginning of the function) redundant?

krshrimali replied:
My bad, fixed now.

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1335,56 +1417,41 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
+    conjugate_mutable_input_if_needed(const_cast<Tensor&>(result), result_is_conj);
+    }
+    namedinference::propagate_names_if_nonempty(
+      const_cast<Tensor&>(result),
```
Do we need this `const_cast`?

ysiraichi replied:
Can we move this name propagation to the `META` function?

krshrimali replied:
Thanks, good suggestions here. These should be resolved now.

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1335,56 +1417,41 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
+      self_or_result.copy_(self);
+    }
+    bool self_is_conj = result.is_conj();
+    conjugate_mutable_input_if_needed(const_cast<Tensor&>(result), self_is_conj);
+    bmm_out_or_baddbmm_(const_cast<Tensor&>(result), batch1.resolve_conj(), batch2.resolve_conj(), beta, alpha, false);
+    conjugate_mutable_input_if_needed(const_cast<Tensor&>(result), self_is_conj);
```
Can we propagate `const` here, instead of casting?

krshrimali replied:
Nice, done!

ysiraichi commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -209,7 +204,6 @@ Tensor& baddbmm_out_cuda_impl(Tensor& result, const Tensor& self, const Tensor&
-  TORCH_CHECK(batch1_sizes[2] == batch2_sizes[1], "batch1 dim 2 must match batch2 dim 1");
 
   if (!result.is_same(self)) {
     result.resize_as_(self);
```
Maybe we can remove this one (since `META` does all the resizing needed).

krshrimali replied:
You are right, this is done now!

ysiraichi reviewed (changes requested):
Hi, @krshrimali. Really nice PR! I feel like these two kernels were quite tricky to port, since they were completely disjoint. Good job! I've left a few observations regarding device-independent checks and name propagation. Let me know if anything is unclear.

ezyang commented:
lmk if I should review now or not

krshrimali commented:
> lmk if I should review now or not

Hi @ezyang - Thanks! I'll address @ysiraichi's review comments and will then request your review and ping you here. :)

krshrimali committed:
Address reviews

krshrimali committed:
Fix merge conflict

krshrimali committed:
NameInference in meta function now

krshrimali committed:
More refactoring

krshrimali commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,86 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+        auto self_ = expand_size(self, {batch1.size(0), batch1.size(1), batch2.size(2)}, "baddbmm");
+        result.resize_(self_->sizes());
+        result.copy_(*self_);
+      }
```
Nice suggestion, thanks Yukio!

krshrimali commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,86 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+
+      IntArrayRef batch1_sizes = batch1.sizes();
+      IntArrayRef batch2_sizes = batch2.sizes();
+      TORCH_CHECK(batch1_sizes[2] == batch2_sizes[1], "batch1 dim 2 must match batch2 dim 1");
```
So I went back and realized that these checks are actually the same for both CUDA and CPU. It's just that they were represented differently.

I've made the checks common, and these are now in `META` function.

krshrimali replied:
Update: checks were common for bmm as well :) In fact a few checks were redundant in CUDA for `bmm`. 

First done here: https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/native/cuda/Blas.cpp#L322-L323

And again here: https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/native/cuda/Blas.cpp#L196-L200

This has been fixed now.

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into structured_kernels/bmm

krshrimali committed:
Checks are common for bmm

krshrimali commented:
This involved some thinking, looks like the checks were self-independent but just represented differently. I've refactored the code, and have addressed the comments (@ysiraichi, thanks for your review and attention to detail).

This should be ready for another review @ysiraichi - if all looks good, then I can request @ezyang's review. :)

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,59 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+    namedinference::propagate_names_if_nonempty(
+      result,
+      namedinference::compute_bmm_outnames(const_cast<Tensor&>(result), self, mat2)
+    );
```
Can't the name inference go in common checks?

krshrimali replied:
Sorry, they definitely should. Will fix this.

krshrimali replied:
Update: fixed in the recent commit. Tests pass locally, should pass with the CI as well.

krshrimali committed:
propagate names in common_checks, minor refactoring

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,54 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+
+  namedinference::propagate_names_if_nonempty(
+    self_or_result,
+    namedinference::compute_bmm_outnames(const_cast<Tensor&>(self_or_result), batch1, batch2)
```
Nit: do you think this `const` can also be propagated to `compute_bmm_outnames`?

krshrimali replied:
Yes, done! :)

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,54 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+
+TORCH_META_FUNC(baddbmm)(const Tensor& self, const Tensor& batch1, const Tensor& batch2, const Scalar& beta, const Scalar& alpha) {
+  auto& result = maybe_get_output(0);
+  if (!result.defined()) {
+    set_output({0}, batch1.options());
+  }
```
Nit: maybe we can get rid of this line, since the next ones do the correct resizing. What do you think?

krshrimali replied:
You are right, this should be fixed now.

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1335,55 +1372,32 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
+(const Tensor & self, const Tensor & batch1, const Tensor & batch2, const Scalar& beta, const Scalar& alpha, const Tensor& result) {
+    bool self_is_conj = result.is_conj();
+    conjugate_mutable_input_if_needed(result, self_is_conj);
+    bmm_out_or_baddbmm_(const_cast<Tensor&>(result), batch1.resolve_conj(), batch2.resolve_conj(), beta, alpha, false);
```
Nit: can we propagate this `const`?

krshrimali replied:
Done! :)

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1335,55 +1372,32 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
+    NoNamesGuard guard;
+    bool result_is_conj = result.is_conj();
+    conjugate_mutable_input_if_needed(result, result_is_conj);
+    bmm_out_or_baddbmm_(const_cast<Tensor&>(result), batch1.resolve_conj(), batch2.resolve_conj(), Scalar(0.0), Scalar(1.0), true);
```
Nit: this one too.

krshrimali replied:
Done! :)

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,54 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+  }
+  if (!result.is_same(self)) {
+    set_output({batch1.size(0), batch1.size(1), batch2.size(2)}, self.options());
+    if (beta.to<c10::complex<double>>() != 0.0) {
+      result.copy_(self);
+    }
```
I think this belongs to the `IMPL` function (it's neither a check nor output metadata setting).

ysiraichi replied:
Then, you would have to replicate it to the CUDA `IMPL` function, too.

krshrimali replied:
The reason it's here is that the size of `self` is modified using `expand_size` (not in-place though, we create another tensor and then use it from there for the checks).

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,54 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+  );
+}
+
+TORCH_META_FUNC(bmm)(const Tensor& self, const Tensor& mat2) {
+    set_output({self.sizes()[0], self.sizes()[1], mat2.sizes()[2]}, self.options());
+    auto& result = maybe_get_output(0);
+    common_checks_baddbmm_bmm(*this, self, mat2, Scalar(0.0), Scalar(1.0), true, result);
+}
```
Nice job! It got brief and straight-forward.

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1335,55 +1372,32 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
-Tensor& baddbmm_out_cpu(const Tensor& self_, const Tensor& batch1, const Tensor& batch2, const Scalar& beta, const Scalar& alpha, Tensor &result) {
-  auto self = expand_size(self_, {batch1.size(0), batch1.size(1), batch2.size(2)}, "baddbmm");
-  result.resize_(self->sizes());
-  result.copy_(*self);
```
I feel like we are losing this copy. But, somehow all the tests are passing.

krshrimali replied:
I thought about it before, and I gave it a thought again.

My observation is, `expand_size(self_, {batch1.size(0), batch1.size(1), batch2.size(2)})` calls `self_.expand` method (if required).

Now the thing is, if `result` tensor's size is already (or if we set explicitly using `set_output`): `{batch1.size(0), batch1.size(1), batch2.size(2)}` - and `self` is not of the required size, then `copy_` will resize the `self` tensor to `result` tensor's size.
 
```python
In [17]: M = torch.randn(10, 1, 5)
In [18]: _out = torch.empty((10, 3, 5))

In [20]: _out.copy_(M).size()
Out[20]: torch.Size([10, 3, 5])
```

And we do this `copy_` in the impl function now (after your suggestion in the review comment above, earlier it was in the meta function) if `result` is not same as `self`.

Now, `result` will be same as `self` if:

1. It's an in-place call (auto-generated code sets `result` as `self` in the meta function itself).
2. `out` is `self` (like: `torch.baddbmm(M, batch1, batch2, out=M)`). -- **I'm not resizing for this case, and that's where it's wrong, please see below**

On upstream:

```python
In [43]: M = torch.randn(10, 1, 5)

In [44]: torch.baddbmm(M, batch1, batch2, out=M).size()
Out[44]: torch.Size([10, 3, 5])
```

On this branch:

```python
In [8]: M = torch.randn(10, 1, 5)

In [9]: torch.baddbmm(M, batch1, batch2, out=M).size()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-4d21e89d2e08> in <module>
----> 1 torch.baddbmm(M, batch1, batch2, out=M).size()

RuntimeError: Expected self_sizes[0] == bs && self_sizes[1] == res_rows && self_sizes[2] == res_cols to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)
```

We need to test this behavior as well in `test_torch.py`. Alright, I'll fix this now.

Thanks for asking this question, @ysiraichi - sorry for the long message. ;)

krshrimali replied:
I'm stuck here:

How would you differentiate these 2 calls (see below) from meta function? (Note: please see the comment above on why we need to differentiate b/w these calls)

```python
>>> M.baddbmm_(batch1, batch2)  # [1] In-Place
>>> torch.baddbmm(M, batch1, batch2, out=M)  # [2] out=self tensor
```

From the auto-generated code for in-place variant in `build/aten/src/ATen/RegisterCPU.cpp`:

```cpp
struct structured_baddbmm_out_cpu_inplace final : public at::native::structured_baddbmm_out_cpu {
     structured_baddbmm_out_cpu_inplace(Tensor& self) : outputs_{std::ref(self)} {}
     // ...
 };
```

As `outputs_` is initialized with `std::ref(self)` so `result.is_same(self)` is always gonna be `true`.

cc: @ezyang @bdhirsh @ysiraichi

Similarly for [2] `out=M` call, `result.is_same(self)` will be `true`. How should we differentiate b/w these calls?

bdhirsh replied:
Hmm, this actually seems more i-line with our out= behavior: for most other out= ops, if you pass in an output tensor with an incorrect size we raise an error instead of transparently resizing the output for you (unless the tensor has zero size).

I'm just not sure if we want a deprecation cycle, since it's not clear to me how commonly this behavior is relied on (if I had to guess though, it seems like a pretty uncommon use case?). I also don't think you have an easy way with structured kernels to differentiate those two cases.

Lezcano replied:
Now, I encountered this behaviour in a test in `test/test_linalg.py` the other day:
https://github.com/pytorch/pytorch/blob/88032d894311e5c0aed8bbc21a4306bc6be4af82/test/test_linalg.py#L4978
Note that the shape is not being passed correctly:
https://github.com/pytorch/pytorch/blob/88032d894311e5c0aed8bbc21a4306bc6be4af82/test/test_linalg.py#L4988-L4995

This will be fixed in https://github.com/pytorch/pytorch/pull/64387 , but it signals that there may be other places where this behaviour is used.

Note that, at the moment, if an incorrectly shaped out is used, a warning is issued. As such, I think that just making that warning into an error may be a reasonable behaviour here.

krshrimali replied:
Thanks for the valuable inputs, @Lezcano, and @bdhirsh. Following the suggestions, we now raise an error for incorrectly shaped `out` (earlier we allowed resizing). However, this behavior is only for `baddbmm`. We resize the output for `bmm`, there is a test that uses incorrect shapes for `out` here: https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L8394.

The only thing that concerns me is the error messages for `inplace` call and `out=M` call. The error with in-place call should have `self_sizes` mentioned instead of `result_sizes`, but that will need distinction b/w the 2 calls.

```python
>>> M = torch.randn(10, 1, 5)  # Incorrect shape
>>> batch1 = torch.randn(10, 3, 4)
>>> batch2 = torch.randn(10, 4, 5)

>>> M.baddbmm_(batch1, batch2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)

>>> torch.baddbmm(M, batch1, batch2, out=M)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)
```

The tests might need another look, which I personally feel can be done in a follow-up PR.

ysiraichi replied:
> How would you differentiate these 2 calls (see below) from meta function? (Note: please see the comment above on why we need to differentiate b/w these calls)

While I agree with @bdhirsh (i.e. you can't), you can resize the output tensors conditionally by using `set_output`. So, instead of writing:

```c++
auto& result = meta.maybe_get_output(0);
if (is_bmm || !result.defined()) {
  meta.set_output({bs, res_rows, res_cols}, batch1.options());
} else {
  const auto result_sizes = result.sizes();
  TORCH_CHECK(result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols);
}
```

You could write:

```c++
// 'set_output' does not resize for in-place calls
meta.set_output({bs, res_rows, res_cols}, batch1.options());
// Error is raised if called from the in-place overload with incorrect self shape
TORCH_CHECK(result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols);
```

This works because the in-place implementation of `set_output` does no resizing!

bdhirsh replied:
@ysiraichi good catch, that seems like it should work (and also raise the warning). It's probably worth quickly confirming that with that code change, we now get the deprecation warning:
```
>>> M = torch.randn(10, 1, 5)  # Incorrect shape
>>> batch1 = torch.randn(10, 3, 4)
>>> batch2 = torch.randn(10, 4, 5)

>>> torch.baddbmm(M, batch1, batch2, out=M)
# (we should see a deprecation warning, because meta.set_output calls at::native::set_output)

krshrimali replied:
This is really cool, @ysiraichi! I wasn't aware that `meta.set_output()` doesn't resize for in-place calls. This helps!

Here is how the use-cases look like with the revised changes:

```python
In [11]: M = torch.randn(10, 1, 5)
In [12]: batch1 = torch.randn(10, 3, 4)
In [13]: batch2 = torch.randn(10, 4, 5)

In [14]: torch.baddbmm(M, batch1, batch2).size()
Out[14]: torch.Size([10, 3, 5])

In [15]: M.baddbmm_(batch1, batch2).size()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-15-ffc510337fae> in <module>
----> 1 M.baddbmm_(batch1, batch2).size()

RuntimeError: Expected result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)

In [16]: torch.baddbmm(M, batch1, batch2, out=M).size()
<ipython-input-16-4d21e89d2e08>:1: UserWarning: An output with one or more elements was resized since it had shape [10, 1, 5], which does not match the required output shape [10, 3, 5].This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at  ../aten/src/ATen/native/Resize.cpp:16.)
  torch.baddbmm(M, batch1, batch2, out=M).size()
Out[16]: torch.Size([10, 3, 5])
```

We do see a deprecation warning with `out=M`, and we also get an error for the in-place call. I've made the required commits, the tests should pass and this PR should hopefully look good now. Welcoming reviews now. :)

ysiraichi reviewed (changes requested):
Hi @krshrimali. That's one nice looking PR. I left some minor comments and some other observations. Check if they make sense, and let me know if there's anything unclear!

krshrimali committed:
Revisions with out behavior, raise errors and remove const_casts

krshrimali committed:
Propagate const

krshrimali committed:
Resize for bmm (out), temporarily?

krshrimali committed:
Lint fixes

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into structured_kernels/bmm

krshrimali committed:
Fix MKL build errors

krshrimali committed:
Propagate const for bmm_out_or_baddbmm_

krshrimali committed:
As per suggestion from @ysiraichi - rely on set_output

ysiraichi commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -57,6 +57,58 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+  meta.set_output({bs, res_rows, res_cols}, batch1.options());
+  const auto result_sizes = result.sizes();
+  // Error is raised if called from in-place overload with incorrect shape
+  TORCH_CHECK(result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols);
```
Nit: an error message here would be nice!

ezyang replied:
You don't have to spell it out individually, ArrayRef has equality defined. This also solves a bug where you can OOB access if the dimensionality differs. Also another one for https://github.com/pytorch/pytorch/issues/57827

ezyang replied:
@ysiraichi maybe you can fix the underlying issue, so we don't have to keep pasting this in? :)

krshrimali replied:
Right, to me figuring out the error message gets tricky here (but I've still made an effort to be as correct as possible, in the recent commit) because an error is raised if:

1. `out` tensor is passed with an incorrect shape.
2. (in-place call), input tensor is passed with an incorrect shape.

I've added an error message in the recent commit, please let me know if it sounds good or if you have any suggestions. However, we'll not reach this error frequently, a lot of times it will fail at `expand_size`:

```python
In [8]: M = torch.randn(1, 3, 2)

In [9]: M.baddbmm_(batch1, batch2)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-a9e665d29b16> in <module>
----> 1 M.baddbmm_(batch1, batch2)

RuntimeError: The expanded size of the tensor (5) must match the existing size (2) at non-singleton dimension 2.  Target sizes: [10, 3, 5].  Tensor sizes: [1, 3, 2]
```

The only time when we'll get this error is for in-place calls, as expected:

```python
In [12]: M = torch.randn(10, 1, 5)

# batch1 size: (10, 3, 4) and batch2 size: (10, 4, 5)
In [13]: M.baddbmm_(batch1, batch2).size()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-13-ffc510337fae> in <module>
----> 1 M.baddbmm_(batch1, batch2).size()

RuntimeError: Expected an output tensor with shape [10, 3, 5] but got shape [10, 1, 5]
```

Hope this error message looks good now. For completeness, I've also added a similar error message for `self_sizes` check in the same function. :)

krshrimali replied:
> You don't have to spell it out individually, ArrayRef has equality defined. This also solves a bug where you can OOB access if the dimensionality differs. Also another one for #57827

Sorry, I just noticed this.  This should be fixed now, defined `output_size` (`IntArrayRef`) since it's used twice.

Also added error messages here:

```python
# batch2 is of incorrect size
In [7]: M.baddbmm_(batch1, batch2)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-a9e665d29b16> in <module>
----> 1 M.baddbmm_(batch1, batch2)

RuntimeError: Expected size for first two dimensions of batch2 tensor to be: [10, 4] but got: [10, 5].
```

[Relevant commit](https://github.com/pytorch/pytorch/pull/64805/commits/86a408220e20f5b8d45accbbbd5b55a74ac30ae2)

ysiraichi reviewed (approved):
LGTM! Nice job, @krshrimali 
@ezyang could you take a look at it?

ezyang commented:
after final review round this looks good to go

krshrimali committed:
Remove out skips for baddbmm and bmm

krshrimali committed:
Improve (actually add) error messages for size checks

krshrimali committed:
Minor lint fix

krshrimali commented:
> after final review round this looks good to go

Thanks, @ezyang! I have added error messages as suggested by @ysiraichi. One thing I missed was removing the out skips from OpInfo, which is now taken care of in: https://github.com/pytorch/pytorch/pull/64805/commits/3ea56a9b015e2b45b8aee5ea0c74c998d19594e7 commit.

Thanks for the patience, and everyone for the reviews.

krshrimali committed:
Apologies, lint failure fix - should be done now

krshrimali committed:
Use IntArrayRef's equality feature, improve error messages

krshrimali committed:
Use int64_t for ArrayRef, Windows tests should pass

krshrimali commented:
Update: working on resolving the test failures.

krshrimali committed:
Use vector instead of ArrayRef

krshrimali committed:
Merge remote-tracking branch 'upstream/master' into structured_kernels/bmm
